.github/workflows/update-snapshots.yml

```diff
@@ -0,0 +1,506 @@
+      - name: Restore node_modules from cache
+        uses: actions/cache@v2
+        with:
+          path: common/temp/pnpm-store
+          key: ${{ runner.os }}-${{ hashFiles('common/config/rush/pnpm-lock.yaml') }}
+      - name: Install rush
+        run: npm install -g @microsoft/rush@5.47.0
+      - name: Install dependencies
+        run: rush install
+      - name: Switch flavor for stable build
+        if: ${{ matrix.flavor == 'stable' }}
+        run: rush switch-flavor:stable
+      - name: Build packages/react-composites browser tests
+        working-directory: ./packages/react-composites
+        run: rushx build:e2e:callwithchat
+      - name: Install packlets used by browser tests
+        run: |
+          cd packages/react-composites
+          rush build -t .
+      - name: Update call with chat snapshot
+        id: update-call-with-chat-snapshots
+        run: rushx test:e2e:callwithchat:update
+        working-directory: ./packages/react-composites
+        env:
+          CONNECTION_STRING: ${{ secrets.CONNECTION_STRING }}
+      - name: Upload snapshot diff
+        if: ${{ always() && steps.update-call-with-chat-snapshots.outcome == 'failure' }}
+        uses: actions/upload-artifact@v2
+        with:
+          name: snapshots
+          path: packages/react-composites/test-results
+      # Check if files have changed
+      # More information see: https://stackoverflow.com/questions/3878624/how-do-i-programmatically-determine-if-there-are-uncommitted-changes
+      - name: Check for snapshot changes
+        id: changescheck
+        run: |
+          if [[ -z $(git status packages -s) ]]
+          then
+            echo "::set-output name=hasChanged::false"
+          else
+            echo "::set-output name=hasChanged::true"
+          fi
+      - name: Push new snapshots, if any
+        if: ${{ steps.changescheck.outputs.hasChanged == 'true' }}
+        # Before pushing changes to origin, merge any intervening changes on the upstream branch.
+        # This allows multiple snapshot update jobs in this action to run concurrently.
+        # - The only files updated locally are UI snapshots
+        # - Each job is responsible for a unique set of UI snapshots
+        # Thus we do not expect any merge conflicts due to the concurrent jobs from this workflows.
+        #
+        # If the UI snapshots are updated in the upstream branch while this job was running
+        # (e.g., due to a merge from the base_ref for a PR), this merge will fail, and the
+        # workflow will have to be triggered again. This is the desired behavior because the workflow
+        # should not silently overwrite updates to the target branch.
+        run: |
+          git add packages/react-composites/*.png
+          git commit -m 'Update packages/react-composites CallWithChatComposite browser test snapshots'
+          git pull origin ${{ needs.get_target_branch.outputs.target }} --no-rebase --no-edit
```

Comment:
instead of commit-pull-push on each action (triggering PR updates each time), another option would be to commit, then have a final job that depends on all the children that does the push

Reply:
Each job runs on a separate node. A single push would require us to use GitHub artifacts API to share commits around. I felt that using git directly was a simpler solution -- to implement, understand and maintain in the future.

The more technologies we bring into a workflow, the more the points of failure as well.

Reply:
To this point though, my approach does introduce an opportunity for flake if multiple jobs finish at the same time.

And at scale, all flake happens frequently enough.

Let me explore the use of artifacts API once as well.

Reply:
Here's an exploration of using artifacts to synchronize patches: https://github.com/prprabhu-ms/scratch-actions/blob/main/.github/workflows/sync.yml

It works well, but It'll add ~150 lines of code to an already complex workflow. 

I think it's worth merging this PR as is for now. If we find that we actually see conflicts / flaky failures due to the concurrent jobs trying to update changes at the same time, we can add the extra complexity.

wdyt?

Reply:
Oh yeah good point on the separate job nodes

Reply:
And yeah I see the artifact method will definitely add even more complexity to an already enormous yml file... gosh yml needs a *proper* way to have re-usable local code